### PR TITLE
CB-12288 Use LB frontend even if LB is not registered with PEM

### PR DESCRIPTION
**Pull request by yg-gd**
tested locally by standing up stack, now uses aws elb address for CM URL
Screen Shot 2021-05-11 at 16 32 50

### CODE REVIEW

File: core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:165]: how could this happen? I don't think this is a valid check and won't be better then a nullpointer on the next line

File: core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:175]: how could this happen?

**Review by lacikaaa — CHANGES_REQUESTED** [review]
please refactor your code as it's hard to follow and not really readable
also your unit tests doesn't cover every case
thanks

File: core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java
Comment (hreeve-cloudera) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:147]: I'm not entirely sure what changing the original stream() to this iterator adds to the logic. It looks like what it's doing is iterating over the set of load balancers to get the PUBLIC load balancer. But that's already what the stream() was doing:
```
LoadBalancer loadBalancer = loadBalancers.stream()
    .filter(lb -> LoadBalancerType.PUBLIC.equals(lb.getType()))
    .findAny()
    .orElseGet(() -> loadBalancers.stream().findAny().orElse(null));
```
That'll look at the set, select the PUBLIC load balancer if one exists in the set, and select the first load balancer in the set if there isn't a PUBLIC one. So I think you can just take the load balancer is selects here and pass that to `getBestAddressable`.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java
Comment (hreeve-cloudera) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:155]: This if statement will only be hit if we have two load balancers in the set that are the same type. Functionally, that shouldn't happen. The load balancers in the set will always have different types, because we only create new load balancer objects if there isn't one of the desired type already saved. (This class also has does the initial load balancer creation, and that check is done in the `createLoadBalancerIfNotExists` method.) I understand what you were doing with this clause, but it's not a valid case due to how the code is written.

Comment (yg-gd) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:155]: This was a piece of information I was missing, Is it correct to say there will only ever be at most 1 Public and 1 Private load balancer returned for a given stackId?

Comment (hreeve-cloudera) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:155]: Correct.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java
Comment (yg-gd) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:156]: I believe all of these could be null, if the loadBalancer had just been initialized before any load balance metadata had been applied to it.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java
Comment (hreeve-cloudera) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:143]: Unresolving this because Mihaly has a point. This is only called when `findAny()` find a valid element, so when could it be null? I understand the caution of wanting to check that it's not, but I think a better path would be to make sure we don't call this method will a null argument.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java
Comment (hreeve-cloudera) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:153]: @lacikaaa That could theoretically happen if the user is looking in the datalake UI while the stack is still spinning up. When the UI tries to get the user facing FQDN for the CM server, it'll land here. If we land here before the metadata has been collected, we just want to return null as an indication that we don't have the information yet.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java
Comment (yg-gd) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java]: That's a good point! I think it's also probably valid to return null in that case, which can signal to anyone else using the method that they should use a different load balancer.

**Comment by hreeve-cloudera** [comment]
@yg-gd The change looks good for me. When merging things into CB, you should squash all your commits in the PR into a single commit before merging, and when you merge choose the "Rebase and merge" option. We'll still need to wait on @lacikaaa approval since he requested changes, but I'll add mine after you do the squash.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java
Comment (lacikaaa) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:155]: I was trying to point out a bit indirect way that this code is not reachable. neither this line nor the first if is reachable. if it is please cover with unit tests, as these are not covered yet. If these can't happen why we have these?
```
        if (StringUtils.isNotBlank(lb.getFqdn())) {
            return lb.getFqdn();
        } else if (StringUtils.isNotBlank(lb.getDns())) {
            return lb.getDns();
        } else {
            return lb.getIp();
        }
```
this should be sufficient if you replace the filters from `notEmpty` to `notBlank`

File: core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:135]: it's much better than previous, I would just move out the 2 methods to make things a bit more nice:
```
    public String getLoadBalancerUserFacingFQDN(Long stackId) {
        Set<LoadBalancer> loadBalancers = loadBalancerPersistenceService.findByStackId(stackId).stream()
                .filter(lb -> StringUtils.isNotBlank(lb.getDns()) || StringUtils.isNotBlank(lb.getIp()) || StringUtils.isNotBlank(lb.getFqdn()))
                .collect(Collectors.toSet());
        return findPublicLbName(loadBalancers)
                .orElseGet(() -> findPrivateLbNameOrNull(loadBalancers));
    }

File: core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java
Comment (hreeve-cloudera) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:155]: The only difference between `isNotBlank` and `isNotEmpty` is `isNotBlank` will return false if the string contains only whitespace, while `isNotEmpty` will return true in that case. They both return false if the string is null or empty.

`testGetLoadBalancerUserFacingFQDN` verifies the first if statement can be reached, `testGetLoadBalancerUserFacingFQDNPublicMissingFQDN` verifies the else/if can be reached, `testGetLoadBalancerUserFacingFQDNIPOnlyMissingFQDN` verifies the final else can be reached, and `testGetLoadBalancerUserFacingFQDNNoFqdnSet` verifies that if none of those are hit then we return null at the end of the method.

Can you explain why you think the code is unreachable?

Comment (lacikaaa) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:155]: thanks, I know the difference. I just wanted to make sure the filter in `getLoadBalancerUserFacingFQDN` is in sync with the conditions here.
first `if` is unreachable because this will never return `null`:
`loadBalancerPersistenceService.findByStackId(stackId)`

the last null is unreachable because of the filter, which filters out all the LB which doesn't have at least one of these field set.

Comment (hreeve-cloudera) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:155]: Ah, I understand now. Thanks for clarifying. Since we're returning null in to different places, the unit test isn't accurately verifying which null return statement is being hit, and so isn't verifying what it's meant to. Thanks for that catch!

@yg-gd to that end I think things would be cleaner if you only had a single null return statement. It would also be good to have a unit test case where the LB exists but has null fqdn, dns, and ip, and a separate test case for when there are no load balancers in the stack at all. That's a test case I missed the first go-round, but it'd be good to have.

Comment (yg-gd) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java]: I took @lacikaaa's point In the case where all 3 are null, then it's safe to return the IP without checking since it's the same as returning null. I've left the initial null check in case this is used in the future in some other context where it's possible that lb is null.

Comment (lacikaaa) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java]: my opinion still: you should drop that check. it's a private method called from one other. this is unused unnecessary code. if we would code like this, we could have tons of null check in the beginning of all methods. the best is to avoid nulls as soon as possible.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:146]: `OrNull` as it can return null if it's not found

File: core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java
Comment (yg-gd) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java]: A compiler is able to infer that it's not possible for lb to be null at this point and will optimize away this check regardless. What remains is whether or not if it's helpful for humans reading the code to understand that inference.

**Review by lacikaaa — APPROVED** [review]

**Pull request merged**
